[CSS] Tokenizer should distinguish newline from whitespace

zakariaridouh opened the pull request:
#### de28c5277420b12db5bfcc0483ec65950b7bc53d
<pre>
[CSS] Tokenizer should distinguish newline from whitespace
<a href="https://bugs.webkit.org/show_bug.cgi?id=276431">https://bugs.webkit.org/show_bug.cgi?id=276431</a>
<a href="https://rdar.apple.com/131474098">rdar://131474098</a>

Reviewed by Darin Adler.

Preserve newlines in Custom CSS Properties. Write unit tests to test the CSSTokenizer and the CSSParser.

    * LayoutTests/fast/css/test-at-supports-newline-expected.txt: Added.
    * LayoutTests/fast/css/test-at-supports-newline.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-syntax/custom-property-rule-ambiguity-expected.txt:
    * Source/WebCore/Headers.cmake:
    * Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    * Source/WebCore/css/CSSValue.h:
    * Source/WebCore/css/calc/CSSCalcExpressionNodeParser.cpp:
    (WebCore::CSSCalcExpressionNodeParser::parseCalcFunction):
    (WebCore::CSSCalcExpressionNodeParser::parseValue):
    (WebCore::CSSCalcExpressionNodeParser::parseCalcProduct):
    (WebCore::CSSCalcExpressionNodeParser::parseCalcSum):
    * Source/WebCore/css/parser/CSSParserImpl.cpp:
    (WebCore::CSSParserImpl::consumeRuleList):
    (WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
    (WebCore::CSSParserImpl::consumeBlockContent):
    * Source/WebCore/css/parser/CSSParserToken.cpp:
    (WebCore::CSSParserToken::CSSParserToken):
    (WebCore::CSSParserToken::hasStringBacking const):
    (WebCore::CSSParserToken::operator== const):
    (WebCore::CSSParserToken::serialize const):
    * Source/WebCore/css/parser/CSSParserToken.h:
    * Source/WebCore/css/parser/CSSParserTokenRange.cpp:
    (WebCore::CSSParserTokenRange::consumeBlockCheckingForEditability):
    (WebCore::CSSParserTokenRange::trimTrailingWhitespace):
    * Source/WebCore/css/parser/CSSParserTokenRange.h:
    (WebCore::CSSParserTokenRange::consumeWhitespace):
    * Source/WebCore/css/parser/CSSSelectorParser.cpp:
    (WebCore::CSSSelectorParser::consumeCombinator):
    * Source/WebCore/css/parser/CSSSupportsParser.cpp:
    (WebCore::CSSSupportsParser::consumeCondition):
    (WebCore::CSSSupportsParser::consumeNegation):
    * Source/WebCore/css/parser/CSSTokenizer.cpp:
    (WebCore::CSSTokenizer::preprocessString):
    (WebCore::CSSTokenizer::isWhitespace):
    (WebCore::CSSTokenizer::isNewline):
    (WebCore::CSSTokenizer::newline):
    (WebCore::twoCharsAreValidEscape):
    (WebCore::CSSTokenizer::whitespace):
    (WebCore::CSSTokenizer::leftParenthesis):
    (WebCore::CSSTokenizer::rightParenthesis):
    (WebCore::CSSTokenizer::leftBracket):
    (WebCore::CSSTokenizer::rightBracket):
    (WebCore::CSSTokenizer::leftBrace):
    (WebCore::CSSTokenizer::rightBrace):
    (WebCore::CSSTokenizer::comma):
    (WebCore::CSSTokenizer::colon):
    (WebCore::CSSTokenizer::semiColon):
    (WebCore::CSSTokenizer::endOfFile):
    (WebCore::CSSTokenizer::consumeStringTokenUntil):
    (WebCore::CSSTokenizer::consumeEscape):
    (WebCore::isNewLine): Deleted.
    (WebCore::CSSTokenizer::whiteSpace): Deleted.
    * Source/WebCore/css/parser/CSSTokenizer.h:
    * Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
    (WebCore::CSSTokenizerInputStream::advanceUntilNonWhitespace):
    (WebCore::CSSTokenizerInputStream::advanceUntilNewlineOrNonWhitespace):
    * Source/WebCore/css/parser/CSSTokenizerInputStream.h:
    * Source/WebCore/css/parser/SizesCalcParser.cpp:
    (WebCore::SizesCalcParser::calcToReversePolishNotation):
    * Source/WebCore/html/track/WebVTTTokenizer.cpp:
    * Source/WebCore/platform/graphics/Color.h:
    (WebCore::Color::Color):
    * Source/WebCore/platform/text/SegmentedString.h:
    * Tools/TestWebKitAPI/CMakeLists.txt:
    * Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
    * Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
    (TestWebKitAPI::TEST(CSSParser, ParseColorInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseColorWithNewlineAndWhitespacesInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseCustomPropertyWithNewlineInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseCustomPropertyWithNewlineAndWhitespacesInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseCustomPropertyWithNewlineBetweenIdentInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseColorPropertyWithNewlineBetweenIdentInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseTextTransformPropertyWithNewlineBetweenTwoIdentInput)):
    (TestWebKitAPI::TEST(CSSPropertyParser, GridTrackLimits)):
    (TestWebKitAPI::TEST(CSSPropertyParserTest, GridTrackLimits)): Deleted.
    * Tools/TestWebKitAPI/Tests/WebCore/CSSTokenizerTests.cpp: Added.
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeWhitespaceLikeInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeIdentAndCommaInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeMultipleWhitespaceInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeNewlineTokenOnlyInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeNewlineAndWhitespaceInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeNewlineWhitespaceAndIdentInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeURLWithProperInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeURLWithNewlineWhitespaceInput)):

Canonical link: <a href="https://commits.webkit.org/282084@main">https://commits.webkit.org/282084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d45e52849841d2c3d50c446e4943da578dbfd590

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61983 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41337 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65963 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12528 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64102 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12838 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49914 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8642 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38370 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30747 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35012 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10906 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11459 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56813 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11210 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67691 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5926 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10946 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57292 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5952 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53619 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57539 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13783 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4851 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37137 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38221 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39317 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37966 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/ASCIICType.h:
```diff
@@ -138,6 +139,11 @@ template<typename CharacterType> constexpr bool isTabOrSpace(CharacterType chara
+template<typename CharacterType> constexpr bool isNewline(CharacterType character)
+{
+    return character == '\n' || character == '\r' || character == '\f';
+}
```
I’m not sure that “newline” is the right name for this concept. The `'\n'` character is named newline. Just because CSS tokenization calls all three of these “newline” doesn’t mean that the right name for it here in this lower level file of fundamental character classes is `isNewline`. I think a longer name for this is probably needed to avoid ambiguity. We might even want to leave it out of `ASCIICType.h` and put it somewhere else if it might be CSS-specific.

zakariaridouh replied:
Yup, I've removed it from `ASCIICType.h` and now using the isNewline defined in `CSSTokenizer.cpp` when needed, as it is CSS-specific.

darinadler commented on Source/WebCore/css/CSSCustomPropertyValue.h:
```diff
@@ -88,7 +88,7 @@ class CSSCustomPropertyValue final : public CSSValue {
     }
 
-    String customCSSText() const;
+    WEBCORE_EXPORT String customCSSText() const;
```
This is wrong. For the test we should call `cssText`, which will call this.

zakariaridouh replied:
Fixed. Wasn't previously aware of `cssText()` which then called `customCSSText()` for the appropriate child class.

darinadler commented on Source/WebCore/css/parser/CSSParserTokenRange.h:
```diff
@@ -85,13 +85,13 @@ class CSSParserTokenRange {
     void consumeComponentValue();
 
-    void consumeWhitespace()
+    void consumeWhitespaceLikeTokens()
```
This set of renames seems unnecessary. I understand that the whitespace token now doesn’t include newlines. But I think that this function and the related ones can include both whitespace and newline without a name change. For example, `isASCIIWhitespace` includes newlines. Just because we are using a separate token for newlines that doesn’t mean that newlines have become “not whitespace“. If we want the naming to be perfectly consistent, I suggest renaming `WhitespaceToken` to `NonNewlineWhitespaceToken`.

This pull request will be a lot smaller if we don’t insist on referring to whitespace including newlines by a new name, and there is no compelling reason to do that.

darinadler commented on Source/WebCore/css/parser/CSSParserTokenRange.h:
```diff
@@ -85,13 +85,13 @@ class CSSParserTokenRange {
+    void consumeWhitespaceLikeTokens()
     {
-        while (peek().type() == WhitespaceToken)
+        while (peek().type() == WhitespaceToken || peek().type() == NewlineToken)
```
Might be helpful to add an `isWhitespace` function that takes a type instead of listing the two different types in each place we have to add this.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParserTest.cpp:
```diff
@@ -0,0 +1,146 @@
+
+using namespace WebCore;
+
+TEST(CSSParserTest, ParseColorInput)
```
Why rename this to add the word "Test" to the end? We don’t do that with most of the other test files, all of which are named after the thing they are testing without the word "Test". We don’t need to inaugurate a new style for these tests, do we?

darinadler commented on Source/WebCore/css/parser/SizesCalcParser.cpp:
```diff
@@ -1,5 +1,5 @@
 // Copyright 2014-2017 The Chromium Authors. All rights reserved.
-// Copyright (C) 2016 Apple Inc. All rights reserved.
+// Copyright (C) 2024 Apple Inc. All rights reserved.
```
Why are you removing the older date? That’s not correct. Most likely should write 2016-2024.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -45,11 +46,31 @@ void CSSTokenizerInputStream::advanceUntilNonWhitespace()
+        while (m_offset < m_stringLength && (isASCIIWhitespace(characters[m_offset]) || isNewline(characters[m_offset])))
             ++m_offset;
     }
 }
```
The `isASCIIWhitespace` function already returns `true` for every character that `isNewline` returns `true` for, so there is no need to add the `isNewline` check here.

zakariaridouh replied:
Fixed. I originally added in the `|| isNewline(characters[m_offset])` to make it more clear in the code, but was unnecessary.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.h:
```diff
@@ -85,6 +85,7 @@ class CSSTokenizerInputStream {
     }
 
     void advanceUntilNonWhitespace();
+    void advanceUntilNonWhitespaceOrNewline();
```
This is the wrong naming. Newlines are whitespace, so the `advanceUntilNonWhitespace` function should not change behavior. The name `advanceUntilNonWhitespaceOrNewline` is ambiguous. Is it "until non-whitespace or newline" or "until non-(whitespace or newline)"? You can see that the normal English grammar sense is not what the function does. The new function should be `advanceUntilNewlineOrNonWhitespace`, and we should update call sites as needed.

darinadler reviewed (changes requested):
Thanks for taking this on. Good change. I have some ideas about how to do it differently.

darinadler commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -532,14 +532,15 @@ bool CSSParserToken::hasStringBacking() const
     case SubstringMatchToken:
     case SuffixMatchToken:
-    case WhitespaceToken:
+    case NonNewlineWhitespaceToken:
```
This list was sorted alphabetically.

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.cpp:
```diff
@@ -140,16 +140,26 @@ unsigned CSSTokenizer::tokenCount()
     return (cc == '\r' || cc == '\n' || cc == '\f');
 }
 
+CSSParserToken CSSTokenizer::newLine(UChar /*cc*/)
```
The commented out `cc` doesn’t add anything. Please just omit it. But also why is this function named “new line”, two words?

zakariaridouh replied:
Yup, but I noticed other similar codepoint functions in the code had the commented out `cc` so I added it for clarity, but I agree it's unnecessary. Removed.

eg: 
```
CSSParserToken CSSTokenizer::colon(UChar /*cc*/)
{
    return CSSParserToken(ColonToken);
}

CSSParserToken CSSTokenizer::semiColon(UChar /*cc*/)
{
    return CSSParserToken(SemicolonToken);
}
```

darinadler replied:
Yes, this is likely remnants of a difference in coding style from the Chromium contributors since this tokenizer was originally done there and imported back to WebKit.

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.h:
```diff
@@ -49,12 +49,15 @@ class CSSTokenizer {
     unsigned tokenCount();
 
+    static bool isNewline(UChar);
+    static bool isWhitespace(CSSParserTokenType);
```
This function would be easier to use as a non-member function and since it takes an argument of a specific type it would be unambiguous and would overload with other similarly named functions. So I would be tempted to make it a non-member. But it’s OK either way.

zakariaridouh replied:
Good point. Made both of them non member functions.

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.h:
```diff
@@ -91,6 +94,7 @@ class CSSTokenizer {
     CSSParserToken blockEnd(CSSParserTokenType, CSSParserTokenType startType);
 
+    CSSParserToken newLine(UChar);
     CSSParserToken whiteSpace(UChar);
```
It’s strange that this is named “white space”, two words.

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.h:
```diff
@@ -91,6 +94,7 @@ class CSSTokenizer {
     CSSParserToken blockStart(CSSParserTokenType blockType, CSSParserTokenType, StringView);
     CSSParserToken blockEnd(CSSParserTokenType, CSSParserTokenType startType);
 
+    CSSParserToken newLine(UChar);
```
It’s strange that this is named “new line”, two words.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -28,6 +28,8 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "config.h"
+#include "CSSTokenizer.h"
```
This is wrong. The top include is the one for the header that corresponds to the `.cpp` file, so `CSSTokenizerInputStream.h` needs to be here. `CSSTokenizer.h` should be included in the separate paragraph below.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -45,11 +47,31 @@ void CSSTokenizerInputStream::advanceUntilNonWhitespace()
         auto characters = m_string->span8();
-        while (m_offset < m_stringLength && isASCIIWhitespace(characters[m_offset]))
+        while (m_offset < m_stringLength && isASCIIWhitespace(characters[m_offset])) {
+            if (CSSTokenizer::isNewline(characters[m_offset]))
```
I am confused now: Given the name of this function, it should not stop at a newline. If we are changing its behavior we should change its name too.

zakariaridouh replied:
Fixed, yup I was got confused as well with the renaming, thanks for catching it.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -45,11 +47,31 @@ void CSSTokenizerInputStream::advanceUntilNonWhitespace()
+    }
+}
+
+void CSSTokenizerInputStream::advanceUntilNewlineOrNonWhitespace()
```
This is the function that should stop at newline.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -45,11 +47,31 @@ void CSSTokenizerInputStream::advanceUntilNonWhitespace()
+    // Using ASCII whitespace here rather than CSS space since we don't do preprocessing
+    if (m_string->is8Bit()) {
+        auto characters = m_string->span8();
+        while (m_offset < m_stringLength && (isASCIIWhitespace(characters[m_offset])))
```
Please remove the parentheses around the call to `isASCIIWhitespace`.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -45,11 +47,31 @@ void CSSTokenizerInputStream::advanceUntilNonWhitespace()
     } else {
         auto characters = m_string->span16();
-        while (m_offset < m_stringLength && isASCIIWhitespace(characters[m_offset]))
+        while (m_offset < m_stringLength && (isASCIIWhitespace(characters[m_offset])))
```
Please remove the parentheses around the call to `isASCIIWhitespace`.

darinadler commented on Source/WebCore/platform/text/SegmentedString.h:
```diff
@@ -27,7 +27,7 @@ namespace WebCore {
 // FIXME: This should not start with "k".
 // FIXME: This is a shared tokenizer concept, not a SegmentedString concept, but this is the only common header for now.
-constexpr LChar kEndOfFileMarker = 0;
+// constexpr LChar kEndOfFileMarker = 0;
```
What’s going on here? If we should remove it please do, but don’t check it in commented out.

darinadler commented on Source/WebCore/xml/parser/XMLDocumentParser.cpp:
```diff
@@ -23,6 +23,8 @@
  * Boston, MA 02110-1301, USA.
  */
 
+#pragma once
```
This is wrong. This `pragma` should only be used in header files.

darinadler commented on Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
```diff
@@ -25,6 +25,8 @@
  * Boston, MA 02110-1301, USA.
  */
 
+#pragma once
```
This is wrong. This `pragma` should only be used in header files.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/CSSTokenizerTest.cpp:
```diff
@@ -0,0 +1,110 @@
+    auto TokenRange = tokenizer.tokenRange();
+    auto TokenRangeSpan = TokenRange.span();
+
+    EXPECT_EQ(TokenRangeSpan.size(), 7ul);
```
These are all backwards. Expected values come first in this macro. If the two values don’t match, the macro logs the value of the second one, and that should be the thing we are checking.

Also, it’s not good to code in the assumption that `size_t` is `unsigned long`. So this should either be `7U` if the width of the integer doesn’t matter or `(size_t)7` or a named constant of type `size_t`.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/CSSTokenizerTest.cpp:
```diff
@@ -0,0 +1,110 @@
+TEST(CSSTokenizerTest, TokenizeWhitespaceLikeInput)
+{
+    CSSTokenizer tokenizer("foo           \n\n\n bar"_s);
+    auto TokenRange = tokenizer.tokenRange();
```
WebKit coding style doesn’t capitalize the name of a local variable. So this should just be `tokenRange`.

darinadler commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -751,12 +752,17 @@ void CSSParserToken::serialize(StringBuilder& builder, const CSSParserToken* nex
             builder.append(' ');
         break;
     }
+    case NewlineToken: {
```
Don't need braces here. We leave them off unless they are needed and there are no local variables that need to be limited in scope.

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.cpp:
```diff
@@ -140,16 +140,26 @@ unsigned CSSTokenizer::tokenCount()
+    return type == NonNewlineWhitespaceToken || type == NewlineToken;
+}
+
+bool isNewline(UChar cc)
```
What I said about overloading and `isWhitespace` doesn't apply to this function, which takes a `UChar`. Nothing in the name of this function or the types of its arguments makes it clear that it's the CSS tokenizer definition of newline, not a globally correct version. So it would be better to leave this one as a static member function or give it a slightly longer name.

zakariaridouh replied:
Moved it back into `CSSTokenizer` and made it a static function.

zakariaridouh replied:
Also, as per Matthieu's comment below, I also made `isWhitespace` back into a static member function of `CSSTokenizer` to prevent putting a isWhitespace function that checks for tokens in the WebCore namespace, where other functions might be expecting it to check for `' '` or similar, even though it seems self explanatory that its related to the tokenizer because it takes in a `CSSParserTokenType`.

darinadler replied:
I don’t agree with Mathieu on this, but I won’t press the point.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "config.h"
 #include "CSSTokenizerInputStream.h"
 
+#include "CSSTokenizer.h"
```
There should be a blank line after this include.

mdubet commented on Source/WebCore/css/parser/CSSTokenizer.h:
```diff
@@ -131,4 +132,7 @@ class CSSTokenizer {
     CSSTokenizerInputStream m_input;
 };
 
+bool isWhitespace(CSSParserTokenType);
```
We could make those `static` inside `CSSTokenizer` to prevent WebCore "pollution" with a pure CSS (weird) concept.

zakariaridouh replied:
Made both `isNewline` and `isWhitespace` static methods inside of `CSSTokenizer`.

mdubet commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -40,6 +41,26 @@ CSSTokenizerInputStream::CSSTokenizerInputStream(const String& input)
+{
+    // Using ASCII whitespace here rather than CSS space since we don't do preprocessing
+    if (m_string->is8Bit()) {
+        auto characters = m_string->span8();
```
We could factorize those 2 code snippets into a lambda.

mdubet commented on Source/WebCore/platform/text/SegmentedString.h:
```diff
@@ -25,10 +25,6 @@
 
 namespace WebCore {
 
-// FIXME: This should not start with "k".
```
Seems to be a cleanup unrelated to this patch ?

zakariaridouh replied:
Was necessary to fix conflicting redefinitions that resulted from new imports of `CSSTokenizer.h`.

mdubet commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
```diff
@@ -34,6 +36,58 @@ namespace TestWebKitAPI {
 
 using namespace WebCore;
 
+TEST(CSSParser, ParseColorInput)
```
Unit tests 👍

mdubet commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
```diff
@@ -34,6 +36,58 @@ namespace TestWebKitAPI {
+    auto properties = MutableStyleProperties::create();
+
+    ASSERT_TRUE(parser.parseDeclaration(properties, "color:  \n    #ff0000;"_s));
+    RefPtr<CSSValue> value = properties->getPropertyCSSValue(CSSPropertyColor);
```
`auto value` would work here

mdubet commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
```diff
@@ -34,6 +36,58 @@ namespace TestWebKitAPI {
+    ASSERT_TRUE(parser.parseDeclaration(properties, "color: #ff0000;"_s));
+    RefPtr<CSSValue> value = properties->getPropertyCSSValue(CSSPropertyColor);
+
+    ASSERT_TRUE(value.get()->isColor());
```
`EXPECT_TRUE` is more appropriate

mdubet commented on Tools/TestWebKitAPI/Tests/WebCore/CSSTokenizerTest.cpp:
```diff
@@ -0,0 +1,110 @@
+    auto tokenRange = tokenizer.tokenRange();
+    auto tokenRangeSpan = tokenRange.span();
+
+    EXPECT_EQ(tokenRangeSpan.size(), 7ul);
```
`ASSERT_EQ` here (because the following code will crash if it's not true)

mdubet commented on LayoutTests/imported/w3c/web-platform-tests/css/css-syntax/custom-property-rule-ambiguity-expected.txt:
```diff
@@ -1,6 +1,6 @@
 PASS Rule that looks like a custom property declaration is ignored
 PASS Rule that looks like an invalid custom property declaration is ignored
-FAIL Nested rule that looks like a custom property declaration assert_equals: expected "hover { }\n    .b { }" but got "hover { }     .b { }"
+PASS Nested rule that looks like a custom property declaration
```
👍

mdubet commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -751,12 +752,17 @@ void CSSParserToken::serialize(StringBuilder& builder, const CSSParserToken* nex
     }
+    case NewlineToken: {
+        if (mode == SerializationMode::CustomProperty)
+            builder.append('\n');
```
We need to append a space here if it's not a custom property. 
Otherwise we could fail to round-trip it seems :
2 identifiers like `foo\nbar` would serialize to a single identifier `foobar` which is wrong.

mdubet replied:
Maybe we can add a test for this

darinadler commented on Source/WebCore/css/parser/CSSParserTokenRange.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "CSSParserTokenRange.h"
+#include "CSSTokenizer.h"
 
 #include "StyleSheetContents.h"
```
The `CSSTokenizer.h` include should be in the paragraph below, not grouped with the main header, so the blank line should be above it, not below it. But also, the `CSSParserTokenRange.h` header now includes it, so we don’t have to also include it here at all.

darinadler commented on Source/WebCore/css/parser/CSSSupportsParser.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "CSSSupportsParser.h"
+#include "CSSTokenizer.h"
 
 #include "CSSParserImpl.h"
```
The `CSSTokenizer.h` include should be in the paragraph below, not grouped with the main header, so the blank line should be above it, not below it.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -42,18 +44,35 @@ CSSTokenizerInputStream::CSSTokenizerInputStream(const String& input)
+        advance(m_string->span8());
+    else
+        advance(m_string->span16());
+
```
Should omit this blank line.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -42,18 +44,35 @@ CSSTokenizerInputStream::CSSTokenizerInputStream(const String& input)
-        while (m_offset < m_stringLength && isASCIIWhitespace(characters[m_offset]))
-            ++m_offset;
+        advance(m_string->span16());
     }
```
For WebKit coding style, should omit the braces here since these are one-line `if`/`else` blocks.

darinadler commented on Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
```diff
@@ -42,18 +44,35 @@ CSSTokenizerInputStream::CSSTokenizerInputStream(const String& input)
 
 void CSSTokenizerInputStream::advanceUntilNonWhitespace()
 {
-    // Using ASCII whitespace here rather than CSS space since we don't do preprocessing
```
What’s the rationale for removing this comment?

zakariaridouh replied:
Was erroneously removed, I've added it back.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
```diff
@@ -25,15 +26,144 @@
+    auto value = properties->getPropertyCSSValue(CSSPropertyColor).get();
+
+    ASSERT_TRUE(is<CSSValue>(value));
+    Color valueColor({ 255, 0, 0 });
```
Should just be able to use `Color::redColor` instead of this.

darinadler commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -1,7 +1,7 @@
  * (C) 1999-2003 Lars Knoll (knoll@kde.org)
  * (C) 2002-2003 Dirk Mueller (mueller@kde.org)
- * Copyright (C) 2002-2021 Apple Inc. All rights reserved.
+ * Copyright (C) 2002-2024 Apple Inc. All rights reserved.
```
Not sure it’s right to be changing the date here but not otherwise touching the file in this pull request.

darinadler commented on Source/WebCore/css/calc/CSSCalcExpressionNodeParser.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021 Apple Inc. All rights reserved.
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
2021-2024

darinadler commented on Source/WebCore/css/parser/CSSSupportsParser.h:
```diff
@@ -1,5 +1,5 @@
 // Copyright 2015 The Chromium Authors. All rights reserved.
-// Copyright (C) 2016 Apple Inc. All rights reserved.
+// Copyright (C) 2016-2024 Apple Inc. All rights reserved.
```
Not sure it’s right to be changing the date here but not otherwise touching the file in this pull request.

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.cpp:
```diff
@@ -661,7 +668,7 @@ static bool isNonPrintableCodePoint(UChar cc)
 CSSParserToken CSSTokenizer::consumeURLToken()
 {
-    m_input.advanceUntilNonWhitespace();
+    m_input.advanceUntilNewlineOrNonWhitespace();
```
Why is it important to stop at a newline here? Do we have a test case covering that change in behavior?

darinadler commented on Source/WebCore/css/parser/CSSTokenizer.cpp:
```diff
@@ -682,7 +689,7 @@ CSSParserToken CSSTokenizer::consumeURLToken()
 
         if (isASCIIWhitespace(cc)) {
-            m_input.advanceUntilNonWhitespace();
+            m_input.advanceUntilNewlineOrNonWhitespace();
```
This now stops at a newline, which presumably fixes a bug? Do we have a test case covering that change in behavior?

zakariaridouh replied:
Reviewed the CSS Spec[1] and added 2 additional tests in `CSSTokenizerTests.cpp` to validate this. Fixed.

Interestingly enough, I was unable to find any behavior differences (both in any API or WebPlatformTests) nor on any websites when running a local build of `advancingUntilNonWhitespace` vs `advancingUntilNewlineOrNonNewline`.

I believe this is due to how we handle tokens after the first `(` of a URL.

[1] https://drafts.csswg.org/css-syntax/#consume-ident-like-token

darinadler replied:
I guess then that we should not change this code since it doesn't need to be changed. The newer function has a longer name and a slightly more complex implementation. Maybe you already did that: having a little trouble reading the pull request and clearly it doesn't really matter either way.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
```diff
@@ -25,15 +26,143 @@
+
+    ASSERT_TRUE(is<CSSValue>(value));
+
+    Color valueColor({ 255, 0, 0 });
```
`Color::red`

darinadler reviewed (approved):
Should land this soon, don’t need to revise based on all my comments, but I did see a few more things to comment on.

zakariaridouh committed:
[CSS] Tokenizer should distinguish newline from whitespace
https://bugs.webkit.org/show_bug.cgi?id=276431
rdar://131474098

Reviewed by Darin Adler.

Preserve newlines in Custom CSS Properties. Write unit tests to test the CSSTokenizer and the CSSParser.

    * LayoutTests/fast/css/test-at-supports-newline-expected.txt: Added.
    * LayoutTests/fast/css/test-at-supports-newline.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-syntax/custom-property-rule-ambiguity-expected.txt:
    * Source/WebCore/Headers.cmake:
    * Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    * Source/WebCore/css/CSSValue.h:
    * Source/WebCore/css/calc/CSSCalcExpressionNodeParser.cpp:
    (WebCore::CSSCalcExpressionNodeParser::parseCalcFunction):
    (WebCore::CSSCalcExpressionNodeParser::parseValue):
    (WebCore::CSSCalcExpressionNodeParser::parseCalcProduct):
    (WebCore::CSSCalcExpressionNodeParser::parseCalcSum):
    * Source/WebCore/css/parser/CSSParserImpl.cpp:
    (WebCore::CSSParserImpl::consumeRuleList):
    (WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
    (WebCore::CSSParserImpl::consumeBlockContent):
    * Source/WebCore/css/parser/CSSParserToken.cpp:
    (WebCore::CSSParserToken::CSSParserToken):
    (WebCore::CSSParserToken::hasStringBacking const):
    (WebCore::CSSParserToken::operator== const):
    (WebCore::CSSParserToken::serialize const):
    * Source/WebCore/css/parser/CSSParserToken.h:
    * Source/WebCore/css/parser/CSSParserTokenRange.cpp:
    (WebCore::CSSParserTokenRange::consumeBlockCheckingForEditability):
    (WebCore::CSSParserTokenRange::trimTrailingWhitespace):
    * Source/WebCore/css/parser/CSSParserTokenRange.h:
    (WebCore::CSSParserTokenRange::consumeWhitespace):
    * Source/WebCore/css/parser/CSSSelectorParser.cpp:
    (WebCore::CSSSelectorParser::consumeCombinator):
    * Source/WebCore/css/parser/CSSSupportsParser.cpp:
    (WebCore::CSSSupportsParser::consumeCondition):
    (WebCore::CSSSupportsParser::consumeNegation):
    * Source/WebCore/css/parser/CSSTokenizer.cpp:
    (WebCore::CSSTokenizer::preprocessString):
    (WebCore::CSSTokenizer::isWhitespace):
    (WebCore::CSSTokenizer::isNewline):
    (WebCore::CSSTokenizer::newline):
    (WebCore::twoCharsAreValidEscape):
    (WebCore::CSSTokenizer::whitespace):
    (WebCore::CSSTokenizer::leftParenthesis):
    (WebCore::CSSTokenizer::rightParenthesis):
    (WebCore::CSSTokenizer::leftBracket):
    (WebCore::CSSTokenizer::rightBracket):
    (WebCore::CSSTokenizer::leftBrace):
    (WebCore::CSSTokenizer::rightBrace):
    (WebCore::CSSTokenizer::comma):
    (WebCore::CSSTokenizer::colon):
    (WebCore::CSSTokenizer::semiColon):
    (WebCore::CSSTokenizer::endOfFile):
    (WebCore::CSSTokenizer::consumeStringTokenUntil):
    (WebCore::CSSTokenizer::consumeEscape):
    (WebCore::isNewLine): Deleted.
    (WebCore::CSSTokenizer::whiteSpace): Deleted.
    * Source/WebCore/css/parser/CSSTokenizer.h:
    * Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
    (WebCore::CSSTokenizerInputStream::advanceUntilNonWhitespace):
    (WebCore::CSSTokenizerInputStream::advanceUntilNewlineOrNonWhitespace):
    * Source/WebCore/css/parser/CSSTokenizerInputStream.h:
    * Source/WebCore/css/parser/SizesCalcParser.cpp:
    (WebCore::SizesCalcParser::calcToReversePolishNotation):
    * Source/WebCore/html/track/WebVTTTokenizer.cpp:
    * Source/WebCore/platform/graphics/Color.h:
    (WebCore::Color::Color):
    * Source/WebCore/platform/text/SegmentedString.h:
    * Tools/TestWebKitAPI/CMakeLists.txt:
    * Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
    * Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
    (TestWebKitAPI::TEST(CSSParser, ParseColorInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseColorWithNewlineAndWhitespacesInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseCustomPropertyWithNewlineInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseCustomPropertyWithNewlineAndWhitespacesInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseCustomPropertyWithNewlineBetweenIdentInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseColorPropertyWithNewlineBetweenIdentInput)):
    (TestWebKitAPI::TEST(CSSParser, ParseTextTransformPropertyWithNewlineBetweenTwoIdentInput)):
    (TestWebKitAPI::TEST(CSSPropertyParser, GridTrackLimits)):
    (TestWebKitAPI::TEST(CSSPropertyParserTest, GridTrackLimits)): Deleted.
    * Tools/TestWebKitAPI/Tests/WebCore/CSSTokenizerTests.cpp: Added.
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeWhitespaceLikeInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeIdentAndCommaInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeMultipleWhitespaceInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeNewlineTokenOnlyInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeNewlineAndWhitespaceInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeNewlineWhitespaceAndIdentInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeURLWithProperInput)):
    (TestWebKitAPI::TEST(CSSTokenizerTest, TokenizeURLWithNewlineWhitespaceInput)):

Canonical link: https://commits.webkit.org/282084@main
